cache/redis_test.go

```diff
@@ -0,0 +1,491 @@
+							Packages: []types.Package{
+								{
+									Name:       "musl",
+									Version:    "1.1.22-r3",
+									SrcName:    "musl",
+									SrcVersion: "1.1.22-r3",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantKey: "fanal::blob::sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0",
+		},
+		{
+			name:       "no such host",
+			setupRedis: false,
+			args: args{
+				blobID:     "sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0",
+				blobConfig: types.BlobInfo{},
+			},
+			wantErr: "unable to store blob information in Redis cache",
+		},
+		{
+			name:       "malformed blobID",
+			setupRedis: true,
+			args: args{
+				blobID:     "sha256:malformed",
+				blobConfig: types.BlobInfo{},
+			},
+			wantErr: "found non-hex character in hash",
+		},
+	}
+
+	// Set up Redis test server
+	s, err := miniredis.Run()
+	require.NoError(t, err)
+	defer s.Close()
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			addr := s.Addr()
+			if !tt.setupRedis {
+				addr = "dummy:16379"
+			}
+
+			c := cache.NewRedisCache(&redis.Options{
+				Addr: addr,
+			})
+
+			err = c.PutBlob(tt.args.blobID, tt.args.blobConfig)
+			if tt.wantErr != "" {
+				require.NotNil(t, err)
+				assert.Contains(t, err.Error(), tt.wantErr)
+				return
+			} else {
+				assert.NoError(t, err)
+			}
+
+			got, err := s.Get(tt.wantKey)
+			require.NoError(t, err)
+
+			want, err := json.Marshal(tt.args.blobConfig)
+			require.NoError(t, err)
+
+			assert.JSONEq(t, string(want), got)
+		})
+	}
+}
+
+func TestRedisCache_GetArtifact(t *testing.T) {
+	tests := []struct {
+		name       string
+		setupRedis bool
+		artifactID string
+		want       types.ArtifactInfo
+		wantErr    string
+	}{
+		{
+			name:       "happy path",
+			setupRedis: true,
+			artifactID: "sha256:8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf",
+			want: types.ArtifactInfo{
+				SchemaVersion: 1,
+				Architecture:  "amd64",
+				Created:       time.Date(2020, 11, 14, 0, 20, 4, 0, time.UTC),
+				DockerVersion: "19.03.12",
+				OS:            "linux",
+			},
+		},
+		{
+			name:       "malformed JSON",
+			setupRedis: true,
+			artifactID: "sha256:961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4",
+			wantErr:    "JSON unmarshal error",
+		},
+		{
+			name:       "no such host",
+			setupRedis: false,
+			artifactID: "sha256:961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4",
+			wantErr:    "failed to get artifact from the Redis cache",
+		},
+		{
+			name:       "nonexistent key",
+			setupRedis: true,
+			artifactID: "sha256:foo",
+			wantErr:    "artifact (sha256:foo) is missing in Redis cache",
+		},
+	}
+
+	// Set up Redis test server
+	s, err := miniredis.Run()
+	require.NoError(t, err)
+	defer s.Close()
+
+	// Set key/value pairs
+	v := `{"SchemaVersion":1,"Architecture":"amd64","Created":"2020-11-14T00:20:04Z","DockerVersion":"19.03.12","OS":"linux"}`
```

Comment:
If we added AsJson method to ArtifactInfo struc, we could avoid JSON string literals in unit tests:

```go
jsonString, err := ArtifactInfo{}.AsJson()
```

WDYT?

Reply:
I think we need string literals to test `AsJson` anyway.

Reply:
Thanks for the detailed explanation. I was mainly wondering why in some cases there are string literals and structs in other places. Other than that I don't have strong opinions about string literals in unit tests as long as they do the job.

The way I read the code is that you're trying to test 2 things: 1) The usage of Go and Redis client so values are stored as JSON 2) JSON marshalling/unmarshaling

If you testes marshalling/unmarshalling somewhere else (which might be overhead) you could probably unit test Redis code with structs only. But then you'll need another integration test that tests both Redis and JSON stuff.

Reply:
Thanks for summarizing. It totally makes sense. I wanted to test 1) and 2) both. But eventually, I switched to structs. https://github.com/aquasecurity/fanal/pull/143/commits/0d842dfea762f3d8241973e4f93f3869f82a7cdd

Mixing string literals and structs in unit tests might be confusing as you said. In addition, it looks enough to test 1) only as we already know JSON marshaling/unmarshaling well.